btree/btree.c

```diff
@@ -1786,16 +1833,24 @@ static int64_t node_alloc(struct node_op *op, struct td *tree, int size,
 	return nxt_state;
 }
 
-#ifndef __KERNEL__
 static int64_t node_free(struct node_op *op, struct nd *node,
 			 struct m0_be_tx *tx, int nxt)
 {
 	int shift = node->n_type->nt_shift(node);
```

Comment:
Add an assert to make sure that all the records in this node have been removed. This is helpful during debugging.

Reply:
If I add `node_count()` check here , it will call `node_invariant()`. Node invariant has check `node->n_skip_rec_count_check || ergo(h->ff_level > 0, h->ff_used > 0)`. This check will fail for `btree_del_kv_tick() --> case P_FREENODE -->  node_free()`, since it will delete internal nodes with 0 records. There is a task for enhancing invariant, is it okay to put this as part of the same task? This task will make sure to enhance invariant and also add invariants in functions

Reply:
Good thing you reminded me. We need to rethink about the invariant. Some way we need to make the invariant aware of these operations or else the invariant should not be too strict.

Reply:
Yes, we need to think about it because current invariants might not catching bugs. In my opinion we should have strict invariants for development phase. May be for development/verification phase we can have some invariants (like function specific) under `M0_PRE_EX`, which we can remove later and some invariants under `M0_PRE` which we can keep.

Reply:
yes we can have M0_PRE_EX but we still need to find how can the invariant inspect the node condition and apply appropriate checks instead of taking hints given  by the caller.